File: DL/python/send_images_cifar.py
"""
send_images_cifar.py: sends labeled CIFAR10 images encoded as a string to an inferencing program

Usage: python3 send_images_cifar.py [-h] [-i IMAGESPERSEC] [-t TOTALIMAGES] | nc <dest IP address>  <dest port>
optional arguments:
  -h, --help            show this help message and exit
  -i IMAGESPERSEC, --imagesPerSec IMAGESPERSEC
  -t TOTALIMAGES, --totalImages TOTALIMAGES
  -s, --subtractMean

CIFAR10 dataset from https://www.cs.toronto.edu/~kriz/cifar.html
(Learning Multiple Layers of Features from Tiny Images, Alex Krizhevsky, 2009, https://www.cs.toronto.edu/~kriz/learning-features-2009-TR.pdf)

Copyright (c) 2019 VMware, Inc.

This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in compliance with the Apache 2.0 License.

This product may include a number of subcomponents with separate copyright notices and license terms. Your use of these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE file.
"""

import argparse
import random
from time import time, gmtime, strftime, sleep, monotonic
import sys
from io import StringIO
import numpy as np
from math import exp
from keras.datasets import cifar10

parser = argparse.ArgumentParser(description='Send CIFAR10 images encoded as strings')
parser.add_argument("-i", "--imagesPerSec", type=int, dest="imagesPerSec", default=10)
parser.add_argument("-t", "--totalImages", type=int, dest="totalImages", default=100)
parser.add_argument("-s", "--subtractMean", action="store_true", dest="subtractMean")
args = parser.parse_args()
images_per_second=args.imagesPerSec; total_images=args.totalImages; subtract_mean=args.subtractMean

def accurate_wait(wait_in_seconds):
    waitUntil = monotonic() +  wait_in_seconds
    while (waitUntil > monotonic()):
      pass

print("%sZ: Loading and normalizing the CIFAR10 data" % (strftime("%Y-%m-%dT%H:%M:%S", gmtime())), file=sys.stderr)
(train_images, train_labels), (test_images, test_labels) = cifar10.load_data()
n_images = test_images.shape[0]
n_labels = test_labels.shape[0]

# Normalize data.
train_images = train_images.astype('float32') / 255
test_images = test_images.astype('float32') / 255

if subtract_mean:
    train_mean = np.mean(train_images, axis=0)
    test_images -= train_mean

# First, write labeled, unraveled images to a list
labeled_images = []
for i in range(n_images):
  string = StringIO()
  np.savetxt(string, test_images[i].ravel().reshape(1,3072), fmt='%f') # 3072 = 32x32x3
  # Insert (single character) label in front of string, cut final '\n' from string
  labeled_images.append(str(test_labels.item(i)) + string.getvalue()[:-1])

print("%sZ: Sending %d images per second for a total of %d images" % (strftime("%Y-%m-%dT%H:%M:%S", gmtime()), images_per_second, total_images), file=sys.stderr, end='')
if subtract_mean:
  print(" with pixel mean subtracted", file=sys.stderr)
else:
  print("", file=sys.stderr)

for i in range(total_images):
  print(labeled_images[i%n_images])
  sys.stdout.flush()
  # Use lognormal distribution to generate a positive random wait time with mean determined from images_per_second and long tail
  mean_wait = float(1.0/images_per_second)
  # Set standard deviation to half the mean_wait
  std_dev = mean_wait/2.0
  fudge_factor = .7 # Needed to reduce wait time to compensate for computation/network time - set empirically
  accurate_wait(fudge_factor*mean_wait*random.lognormvariate(mean_wait,std_dev)/exp(mean_wait + std_dev**2/2))
  if (((i+1) % images_per_second == 0) or (i == total_images-1)):
    print("%sZ: %d images sent" % (strftime("%Y-%m-%dT%H:%M:%S", gmtime()), i+1), file=sys.stderr)

print("")  # Indicate end of send
print("%sZ: Image stream ended" % (strftime("%Y-%m-%dT%H:%M:%S", gmtime())), file=sys.stderr)

File: DL/scala/src/main/scala/com/iot-analytics-benchmark/send_images_cifar.scala
/*
send_images_cifar.scala: sends CIFAR10 images encoded as a string to a Spark Streaming inference program - Scala version

Usage:
  $ scala -J-Xmx128g -cp <path>/scopt_2.11-3.7.1.jar:scala/target/scala-2.11/infer_stream_cifar_2.11-0.0.1.jar com.intel.analytics.bigdl.models.resnet.send_images_cifar \
    <arguments> | nc -lk <port>
OR
  $ java -Xmx128g -cp <path>/scala-library.jar:<path>/scopt_2.11-3.7.1.jar:scala/target/scala-2.11/infer_stream_cifar_2.11-0.0.1.jar \ 
    com.intel.analytics.bigdl.models.resnet.send_images_cifar <arguments> | nc -lk <port>

Arguments:
  -f, --folder <value>        the location of Cifar10 dataset  Default: datasets/cifar-10-batches-bin
  -i, --imagesPerSec <value>  images per second                Default: 10
  -t, --totalImages <value>   total images                     Default: 100

Uses Intel's BigDL library (https://github.com/intel-analytics/BigDL) and CIFAR10 dataset from https://www.cs.toronto.edu/~kriz/cifar.html
  
This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in compliance with the Apache 2.0 License.

This product may include a number of subcomponents with separate copyright notices and license terms. Your use of these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE file.
*/

package com.intel.analytics.bigdl.models.resnet

import scala.collection.mutable.ArrayBuffer
import scala.math._
import java.nio.ByteBuffer
import java.nio.file.{Files, Path, Paths}
import java.time.Instant
import java.util.Random
import scopt.OptionParser

object send_images_cifar {

  // From https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/Utils.scala, with HDFS removed
  case class ByteRecord(data: Array[Byte], label: Float)
  def load(featureFile: String, result : ArrayBuffer[ByteRecord]): Unit = {
    val rowNum = 32
    val colNum = 32
    val imageOffset = rowNum * colNum * 3 + 1
    val channelOffset = rowNum * colNum
    val bufferOffset = 8

    val featureBuffer =
      ByteBuffer.wrap(Files.readAllBytes(Paths.get(featureFile)))
    

    val featureArray = featureBuffer.array()
    val featureCount = featureArray.length / (rowNum * colNum * 3 + 1)

    var i = 0
    while (i < featureCount) {
      val img = new Array[Byte]((rowNum * colNum * 3 + bufferOffset))
      val byteBuffer = ByteBuffer.wrap(img)
      byteBuffer.putInt(rowNum)
      byteBuffer.putInt(colNum)

      val label = featureArray(i * imageOffset).toFloat
      var y = 0
      val start = i * imageOffset + 1
      while (y < rowNum) {
        var x = 0
        while (x < colNum) {
          img((x + y * colNum) * 3 + 2 + bufferOffset) =
            featureArray(start + x + y * colNum)
          img((x + y * colNum) * 3 + 1 + bufferOffset) =
            featureArray(start + x + y * colNum + channelOffset)
          img((x + y * colNum) * 3 + bufferOffset) =
            featureArray(start + x + y * colNum + 2 * channelOffset)
          x += 1
        }
        y += 1
      }
      result.append(ByteRecord(img, label + 1.0f))
      i += 1
    }
  }

  // From https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/Utils.scala
  def loadTest(dataFile: String): Array[ByteRecord] = {
    val result = new ArrayBuffer[ByteRecord]()
    val testFile = dataFile + "/test_batch.bin"
    load(testFile, result)
    result.toArray
  }

  def accurateWait(wait_in_seconds: Double): Unit = {
    val  waitUntil = System.nanoTime() + round(1000000000 * wait_in_seconds)
    while (waitUntil > System.nanoTime()) {}
  }

  def main(args: Array[String]) {

    case class Params(
      folder: String = "datasets/cifar-10-batches-bin",
      imagesPerSec: Int  = 10,
      totalImages: Int = 100
    )

    val parser = new OptionParser[Params]("send_images_cifar") {
      opt[String]('f', "folder")
        .text("the location of Cifar10 dataset")
        .action((x, c) => c.copy(folder = x))
      opt[Int]('i', "imagesPerSec")
        .text("images per second")
        .action((x, c) => c.copy(imagesPerSec = x))
      opt[Int]('t', "totalImages")
        .text("total images")
        .action((x, c) => c.copy(totalImages = x))
     }

    parser.parse(args, Params()).foreach { param =>

      System.err.println("Will send " + param.imagesPerSec + " images per second for a total of " + param.totalImages + " images")

      //  First, write labeled test images to an array 
      val image_string_array = new Array[String](10000)
      val images = loadTest(param.folder)
      for (i <- 0 to images.length-1) {
        val d = images(i).data.mkString(",")
        val l = images(i).label.toString
        val dl = l + "," + d
        image_string_array(i) = dl
      }

      System.err.println("Pausing 15 seconds - start image_stream_cifar")
      Thread.sleep(15000)

      // Use lognormal distribution to generate a positive random wait time with mean determined from imagesPerSecond and long tail
      val rand = new Random
      val mean_wait = 1.0/param.imagesPerSec
      // Set standard deviation to half the mean_wait
      val std_dev = mean_wait/2.0

      val t1 = System.nanoTime

//  Send images from array, wrapping if necessary
      System.err.println(Instant.now() + ": Sending images")
      for (i <- 0 to param.totalImages-1) {
        println(image_string_array(i % image_string_array.length))
        System.out.flush()
        if ((i+1) % param.imagesPerSec == 0 ) System.err.println(Instant.now() + ": " + (i+1) + " images sent")
        // Wait a random time from lognormal distribution with mean mean_wait
        accurateWait(mean_wait * exp(rand.nextGaussian()*std_dev + mean_wait)/exp(mean_wait + pow(std_dev,2)/2))
      }

      val t2 = System.nanoTime

      System.err.println(Instant.now() + ": Sent " + param.totalImages + " images in " + "%.1f".format((t2-t1)/1000000000.0) + " seconds")
    }
  }
}
